Example/CocoaPods-AppStoreConnect-Swift-SDK/Tests/ListAppsTests.swift

```diff
@@ -0,0 +1,79 @@
+            "fields[betaLicenseAgreements]": ListApps.Field.BetaLicenseAgreements.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "filter[bundleId]": bundleIds.joinedByCommas(),
+            "filter[id]": ids.joinedByCommas(),
+            "filter[name]": names.joinedByCommas(),
+            "filter[sku]": skus.joinedByCommas(),
+            "include": ListApps.Relationship.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "limit": "\(appsLimit)",
+            "sort": ListApps.Sorting.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "fields[preReleaseVersions]": ListApps.Field.PreReleaseVersions.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "limit[preReleaseVersions]": "\(preReleaseVersionsLimit)",
+            "fields[betaAppReviewDetails]": ListApps.Field.BetaAppReviewDetails.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "fields[betaAppLocalizations]": ListApps.Field.BetaAppLocalizations.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "fields[builds]": ListApps.Field.Builds.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "fields[betaGroups]": ListApps.Field.BetaGroups.allCases.map({ $0.rawValue }).joinedByCommas(),
+            "limit[builds]": "\(buildsLimit)",
+            "limit[betaGroups]": "\(betaGroupsLimit)",
+            "limit[betaAppLocalizations]": "\(betaAppLocalizationsLimit)"
+        ]
+
+        let listApps = APIEndpoint<AppsResponse>.apps(
+            select: [
+                .apps(ListApps.Field.Apps.allCases),
+                .betaLicenseAgreements(ListApps.Field.BetaLicenseAgreements.allCases),
+                .preReleaseVersions(ListApps.Field.PreReleaseVersions.allCases),
+                .betaAppReviewDetails(ListApps.Field.BetaAppReviewDetails.allCases),
+                .betaAppLocalizations(ListApps.Field.BetaAppLocalizations.allCases),
+                .builds(ListApps.Field.Builds.allCases),
+                .betaGroups(ListApps.Field.BetaGroups.allCases)],
+            filters: [
+                .bundleId(bundleIds),
+                .id(ids),
+                .name(names),
+                .sku(skus)],
+            include: ListApps.Relationship.allCases,
+            sortBy: ListApps.Sorting.allCases,
+            limits: [
+                .apps(appsLimit),
+                .preReleaseVersions(preReleaseVersionsLimit),
+                .builds(buildsLimit),
+                .betaGroups(betaGroupsLimit),
+                .betaAppLocalizations(betaAppLocalizationsLimit)])
+        
+        XCTAssertEqual(listApps.method, .get)
+        XCTAssertEqual(listApps.path, "apps")
+        XCTAssertNotNil(listApps.parameters)
+        if let listAppsParameters = listApps.parameters as? [String: String] {
+            XCTAssertEqual(listAppsParameters.count, expectedParams.count)
+            for (key, value) in expectedParams {
+                XCTAssertEqual(listAppsParameters[key], value)
+            }
+        }
+    }
```

Comment:
Can't we simply match on the outcoming URL absoluteString? That would improve readability a lot and makes it easier to maintain

Reply:
I think testing upon the method, path, and parameters is superior to testing upon the out coming URL absoluteString because: 
- Creation of the URL is not APIEndpoint's job. 
- Order of the query parameters should not matter: `"apps?k1=v1&k2=v2"` vs. `"apps?k2=v2&k1=v1"` 
- If/When switching from Alamofire to URLSession it will still be a valid test.

What do you think ?

Reply:
I still rather prefer a simpler way of testing as we need to maintain the code in the future and we need to test other endpoitns as well.

Writing a test by simply copying the outcome URL is a lot easier than writing all these parameters like this.

The parameters order in the query should be consistent and should not affect this test.

Curious about your thoughts!

Reply:
I would look like this :

```swift
func testListAppsEndpoint() {
    let listApps = APIEndpoint<AppsResponse>.apps(
        select: [
            .apps(ListApps.Field.App.allCases),
            .betaLicenseAgreements(ListApps.Field.BetaLicenseAgreement.allCases),
            .preReleaseVersions(ListApps.Field.PreReleaseVersion.allCases),
            .betaAppReviewDetails(ListApps.Field.BetaAppReviewDetail.allCases),
            .betaAppLocalizations(ListApps.Field.BetaAppLocalization.allCases),
            .builds(ListApps.Field.Build.allCases),
            .betaGroups(ListApps.Field.BetaGroup.allCases)],
        filters: [
            .bundleId(["bundleId"]),
            .id(["id"]),
            .name(["name"]),
            .sku(["sku"])],
        include: ListApps.Relationship.allCases,
        sortBy: ListApps.Sorting.allCases,
        limits: [
            .apps(1),
            .preReleaseVersions(2),
            .builds(3),
            .betaGroups(4),
            .betaAppLocalizations(5)])
        
        let absoluteString = provider.request(for: listApps).request?.url?.absoluteString
        let expected = "https://api.appstoreconnect.apple.com/v1/apps?fields%5Bapps%5D=betaAppLocalizations%2CbetaAppReviewDetail%2CbetaGroups%2CbetaLicenseAgreement%2CbetaTesters%2Cbuilds%2CbundleId%2Cname%2CpreReleaseVersions%2CprimaryLocale%2Csku&fields%5BbetaAppLocalizations%5D=app%2Cdescription%2CfeedbackEmail%2Clocale%2CmarketingUrl%2CprivacyPolicyUrl%2CtvOsPrivacyPolicy&fields%5BbetaAppReviewDetails%5D=app%2CcontactEmail%2CcontactFirstName%2CcontactLastName%2CcontactPhone%2CdemoAccountName%2CdemoAccountPassword%2CdemoAccountRequired%2Cnotes&fields%5BbetaGroups%5D=app%2CbetaTesters%2Cbuilds%2CcreatedDate%2CisInternalGroup%2Cname%2CpublicLink%2CpublicLinkEnabled%2CpublicLinkId%2CpublicLinkLimit%2CpublicLinkLimitEnabled&fields%5BbetaLicenseAgreements%5D=agreementText%2Capp&fields%5Bbuilds%5D=app%2CappEncryptionDeclaration%2CbetaAppReviewSubmission%2CbetaBuildLocalizations%2CbetaGroups%2CbuildBetaDetail%2CexpirationDate%2Cexpired%2CiconAssetToken%2CindividualTesters%2CminOsVersion%2CpreReleaseVersion%2CprocessingState%2CuploadedDate%2CusesNonExemptEncryption%2Cversion&fields%5BpreReleaseVersions%5D=app%2Cbuilds%2Cplatform%2Cversion&filter%5BbundleId%5D=bundleId&filter%5Bid%5D=id&filter%5Bname%5D=name&filter%5Bsku%5D=sku&include=betaAppLocalizations%2CbetaAppReviewDetail%2CbetaGroups%2CbetaLicenseAgreement%2Cbuilds%2CpreReleaseVersions&limit=1&limit%5BbetaAppLocalizations%5D=5&limit%5BbetaGroups%5D=4&limit%5Bbuilds%5D=3&limit%5BpreReleaseVersions%5D=2&sort=%2BbundleId%2C-bundleId%2C%2Bname%2C-name%2C%2Bsku%2C-sku"
        
    XCTAssertEqual(absoluteString, expected)
}
```

Do you prefer that way ?

Reply:
Yes, I do! Although it's not ideal either, it saves us from writing and creating a lot of extra code like this:

```
        let expectedParams: [String: String] = [
            "fields[apps]": ListApps.Field.Apps.allCases.map({ $0.rawValue }).joinedByCommas(),
            "fields[betaLicenseAgreements]": ListApps.Field.BetaLicenseAgreements.allCases.map({ $0.rawValue }).joinedByCommas(),
            "filter[bundleId]": bundleIds.joinedByCommas(),
            "filter[id]": ids.joinedByCommas(),
            "filter[name]": names.joinedByCommas(),
            "filter[sku]": skus.joinedByCommas(),
            "include": ListApps.Relationship.allCases.map({ $0.rawValue }).joinedByCommas(),
            "limit": "\(appsLimit)",
            "sort": ListApps.Sorting.allCases.map({ $0.rawValue }).joinedByCommas(),
            "fields[preReleaseVersions]": ListApps.Field.PreReleaseVersions.allCases.map({ $0.rawValue }).joinedByCommas(),
            "limit[preReleaseVersions]": "\(preReleaseVersionsLimit)",
            "fields[betaAppReviewDetails]": ListApps.Field.BetaAppReviewDetails.allCases.map({ $0.rawValue }).joinedByCommas(),
            "fields[betaAppLocalizations]": ListApps.Field.BetaAppLocalizations.allCases.map({ $0.rawValue }).joinedByCommas(),
            "fields[builds]": ListApps.Field.Builds.allCases.map({ $0.rawValue }).joinedByCommas(),
            "fields[betaGroups]": ListApps.Field.BetaGroups.allCases.map({ $0.rawValue }).joinedByCommas(),
            "limit[builds]": "\(buildsLimit)",
            "limit[betaGroups]": "\(betaGroupsLimit)",
            "limit[betaAppLocalizations]": "\(betaAppLocalizationsLimit)"
        ]
```

Also, this code would contain quite some logic like `joinedByCommas()` which almost requires us to write a test for our tests.

Do you agree with me?

Reply:
Yes I agree about the use of `joinedByCommas`. 
I don't have a better proposition for now, let's go for it.